swim/ping_request_test.go

```diff
@@ -119,3 +123,250 @@ func (s *PingRequestTestSuite) TestTimesOut() {
+// does and the ping reached the target. The indirectPing therefor return
+// that it has successfully reached the target.
+func TestIndirectPing4(t *testing.T) {
+	cont := make(chan bool, 1)
+
+	tnodes := genChannelNodes(t, 4)
+	defer destroyNodes(tnodes...)
+
+	sender, helper1, helper2, target := tnodes[0], tnodes[1], tnodes[2], tnodes[3]
+	bootstrapNodes(t, sender, helper1, helper2, target)
+	waitForConvergence(t, time.Second/2, sender, helper1, helper2, target)
+
+	targetHostPort := target.node.Address()
+
+	helper2.channel.Close()
+	waitForTChannelClosed(sender.channel, helper2.node.Address())
+
+	helper1.node.RegisterListener(onPingRequestReceive(func() {
+		<-cont
+	}))
+
+	sender.node.RegisterListener(onPingRequestComplete(func() {
+		cont <- true
+	}))
+
+	reached, errs := indirectPing(sender.node, targetHostPort, 2, time.Second)
+	assert.True(t, reached, "expected that target is reached")
+	assert.Len(t, errs, 1, "expected one connection error from the helper nodes")
+}
+
+// Test where target is not bootstrapped and responds with not-readies,
+// indirectPing cannot reach the target and returns false.
+func TestIndirectPing5(t *testing.T) {
+	tnodes := genChannelNodes(t, 4)
+	defer destroyNodes(tnodes...)
+
+	// don't boostrap the target
+	sender, helper1, helper2, target := tnodes[0], tnodes[1], tnodes[2], tnodes[3]
+	bootstrapNodes(t, sender, helper1, helper2)
+	waitForConvergence(t, time.Second/2, sender, helper1, helper2)
+
+	// Add an bootstrapped node.
+	targetHostPort := target.node.Address()
+
+	reached, errs := indirectPing(sender.node, targetHostPort, 2, time.Second)
+	assert.False(t, reached, "expected that target is unreachable")
+	assert.Len(t, errs, 0, "expected no errors from the helper nodes")
+}
+
+// Test where target node is unreachable, indirectPing is negative and
+// returns false.
+func TestIndirectPing6(t *testing.T) {
+	tnodes := genChannelNodes(t, 4)
+	defer destroyNodes(tnodes...)
+
+	sender, helper1, helper2, target := tnodes[0], tnodes[1], tnodes[2], tnodes[3]
+	bootstrapNodes(t, sender, helper1, helper2, target)
+	waitForConvergence(t, time.Second/2, sender, helper1, helper2, target)
+
+	targetHostPort := target.node.Address()
+	target.channel.Close()
+	waitForTChannelClosed(sender.channel, target.node.Address())
+
+	reached, errs := indirectPing(sender.node, targetHostPort, 2, time.Second)
+	assert.False(t, reached, "expected that target is not reached")
+	assert.Len(t, errs, 0, "expected no errors from the helper nodes")
+}
+
+// Test where helper nodes are unreachable, indirectPing is inconclusive
+// and returns false to indicate that the target hasn't been reached.
+func TestIndirectPing7(t *testing.T) {
+	tnodes := genChannelNodes(t, 4)
+	defer destroyNodes(tnodes...)
+
+	bootstrapNodes(t, tnodes...)
+	waitForConvergence(t, time.Second/2, tnodes...)
+	sender, helper1, helper2, target := tnodes[0], tnodes[1], tnodes[2], tnodes[3]
```

Comment:
Can we make waitForTChannelClosed part of the testNode type?

```
helper1.closeAndWait(sender.tchannel)
```

I think it could help DRY things up a bit.